[clang][c++20] Fixed ambiguous-reversed-operator for PixelModuleName

smuzaffar opened the pull request:
This PR should fix the clang IBs warnings [a]. 
@makortel @Dr15Jones , I don't know if this is the correct way to fix these warnings ... let me know if there is correct/better way 

[a] https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_14_1_CLANG_X_2024-07-17-1700/CondFormats/SiPixelObjects
```
  src/CondFormats/SiPixelObjects/src/PixelFEDLink.cc:76:16: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'PixelBarrelName' and 'PixelBarrelName') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
    76 |       if (curr == prev) {
      |           ~~~~ ^  ~~~~
src/DataFormats/TrackerCommon/interface/PixelBarrelName.h:65:8: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
   65 |   bool operator==(const PixelModuleName&) const override;
      |        ^
  src/CondFormats/SiPixelObjects/src/PixelFEDLink.cc:81:18: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'PixelBarrelName' and 'PixelBarrelName') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
    81 |       if (!(curr == prev)) {
      |             ~~~~ ^  ~~~~
src/DataFormats/TrackerCommon/interface/PixelBarrelName.h:65:8: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
   65 |   bool operator==(const PixelModuleName&) const override;
      |        ^
  src/CondFormats/SiPixelObjects/src/PixelFEDLink.cc:108:18: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'PixelEndcapName' and 'PixelEndcapName') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
   108 |       if (!(curr == prev))
      |             ~~~~ ^  ~~~~
src/DataFormats/TrackerCommon/interface/PixelEndcapName.h:67:8: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
   67 |   bool operator==(const PixelModuleName&) const override;

```

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"dd24919d06ae853b89cf7f26c24b4d375d71a631":{"files":["DataFormats/TrackerCommon/interface/PixelBarrelName.h","DataFormats/TrackerCommon/interface/PixelEndcapName.h"],"squashed":false,"time":1721296045}},"emoji":{},"last_seen_sha":"dd24919d06ae853b89cf7f26c24b4d375d71a631","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45492/40961

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45492/40961/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45492/40961/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45492/40962

cmsbuild commented:
A new Pull Request was created by  @smuzaffar for master.

It involves the following packages:

- DataFormats/TrackerCommon (**reconstruction**)


@cmsbuild, @jfernan2, @mandrenguyen can you please review it and eventually sign? Thanks.
@JanFSchulte, @VinInn, @VourMa, @gpetruc, @missirol, @mmusich, @mtosi, @rovere this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smuzaffar commented:
please test for CMSSW_14_1_CLANG_X

makortel commented:
Adding this overload is the right way to fix the "ambiguity", but looking at the implementation of `operator==(const PixelModuleName& o)` functions
https://github.com/cms-sw/cmssw/blob/1a77414004a9b7b3c7a656638a2d2600867dd0cf/DataFormats/TrackerCommon/src/PixelBarrelName.cc#L722-L729
https://github.com/cms-sw/cmssw/blob/1a77414004a9b7b3c7a656638a2d2600867dd0cf/DataFormats/TrackerCommon/src/PixelEndcapName.cc#L356-L363
it would be better to have the overloads added in this PR to have the actual implementation of the comparison, and then have the "generic" comparison operator to call these overloads depending on `PixelModuleName::isBarrel()`.

(I somewhat wonder though if the "generic" operator is actually being used after the addition of these overloads)

smuzaffar commented:
@makortel , I have updated this PR according to the changes you suggested in https://github.com/cms-sw/cmssw/pull/45492#issuecomment-2236638814 . I noticed that there were two more classes `PixelBarrelNameUpgrade` and `PixelEndcapNameUpgrade` which are also drived from `PixelModuleName` , so I updated those too.

> (I somewhat wonder though if the "generic" operator is actually being used after the addition of these overloads)

I tried remobing the generic operator but then I get build errors like
```
src/CalibTracker/SiPixelConnectivity/src/PixelToLNKAssociateFromAscii.cc:24:30: error: invalid operands to binary expression ('const PixelModuleName' and 'const PixelModuleName')
   24 |     if ((*(im->first.module) == *roc.module) && (im->first.rocDetId == roc.rocDetId)) {
      |          ~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~
gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../include/c++/12.3.1/system_error:444:3: note: candidate function not viable: no known conversion from 'const PixelModuleName' to 'const error_code' for 1st argument
  444 |   operator==(const error_code& __lhs, const error_code& __rhs) noexcept
```

So I guess the generic operator is needed

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45492/40970

cmsbuild commented:
Pull request #45492 was updated. @civanch, @cmsbuild, @jfernan2, @mandrenguyen, @mdhildreth can you please check and sign again.

makortel commented:
> I tried remobing the generic operator but then I get build errors like
> ...
> So I guess the generic operator is needed

Ok, thanks for the test!

civanch commented:
Not fully understand the problem but would not recommend to have two return statements.
Also suspect that a separate if (!o.isBarrel())  not necessary needed.

smuzaffar commented:
> Not fully understand the problem but would not recommend to have two return statements. 

So you prefer to have something like
```
  return other.isBarrel() && (dynamic_cast<const PixelBarrelNameUpgrade&>(other) == *this);
```

> Also suspect that a separate if (!o.isBarrel()) not necessary needed.
this I do not understand

civanch commented:
Yes, I would simplify the expression in a way you propose.

smuzaffar committed:
c++20: Fixed ambiguous-reversed-operator for PixelModuleName

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45492/40975

cmsbuild commented:
Pull request #45492 was updated. @civanch, @cmsbuild, @jfernan2, @mandrenguyen, @mdhildreth can you please check and sign again.

smuzaffar commented:
please test

smuzaffar commented:
please test for CMSSW_14_1_CLANG_X

cmsbuild commented:
+1

**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-504be8/40491/summary.html
**COMMIT**: 45fe07d45f9de10f14af49914678fc5397f3e65b
**CMSSW**: CMSSW_14_1_X_2024-07-18-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45492/40491/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



**Summary**:
 - You potentially [added 8 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63663/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 167 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63663/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3451730
 - DQMHistoTests: [Total failures: 313](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63663/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3451397
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 206 log files, 170 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63663/triggerResults): no differences found

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-504be8/40492/summary.html
**COMMIT**: 45fe07d45f9de10f14af49914678fc5397f3e65b
**CMSSW**: CMSSW_14_1_CLANG_X_2024-07-18-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45492/40492/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 709 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63664/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 37763 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63664/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3451730
 - DQMHistoTests: [Total failures: 35567](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63664/validateJR.html)
 - DQMHistoTests: Total nulls: 120
 - DQMHistoTests: Total successes: 3416023
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.05 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 140.043 ): -0.008 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 140.063 ): 0.004 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 141.042 ): 0.023 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 141.044 ): 0.043 KiB        JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 141.046 ): -0.012 KiB       JetMET/SUSYDQM
 - Checked 206 log files, 170 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-18-2300+504be8/63664/triggerResults): found differences in 7 / 47 workflows

civanch commented:
@smuzaffar , is the warning fixed? Why we have too many difference in comparisons?

smuzaffar commented:
@civanch , yes warnings are gone https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-504be8/40492/build-logs/.

comparison diff here https://github.com/cms-sw/cmssw/pull/45492#issuecomment-2239582084 are due to CLANG vs Default IB, so these are expected. 
Comparison differences here https://github.com/cms-sw/cmssw/pull/45492#issuecomment-2238973448 are due to dd4hep workflow

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @mandrenguyen, @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
